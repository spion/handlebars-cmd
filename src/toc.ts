import * as fs from 'fs';
import * as path from 'path'

type Entry = {
  name: string;
  link: string;
  children?: Entry[];
};

let ignores = /node_modules|^\.|_sidebar|_docsify/;
let isDoc = /.md$/;

export class TOC {

private niceName(name: string) {
  let splitName = name.split('-');
  if (Number.isNaN(Number(splitName[0]))) return splitName.join(' ');
  return splitName.slice(1).join(' ');
}
private buildTree(dirPath: string, name = '', dirLink = ''): Entry {
  let children: Entry[] = [];
  for (let fileName of fs.readdirSync(dirPath)) {
    if (ignores.test(fileName)) continue;

    let fileLink = dirLink + '/' + fileName;
    let filePath = path.join(dirPath, fileName);
    if (fs.statSync(filePath).isDirectory()) {
      let sub = this.buildTree(filePath, fileName, fileLink);
      if (sub.children != null && sub.children.length > 0)
        children.push(sub);
    } else if (isDoc.test(fileName)) {
      children.push({ name: fileName, link: fileLink });
    }
  }

  return { name, children, link: dirLink };
}
private renderToMd(tree: Entry, linkDir = false): string {
  if (!tree.children) {
    return `- [${this.niceName(path.basename(tree.name, '.md'))}](${tree.link.replace(/ /g, '%20')})`;
  } else {
    let fileNames = new Set(tree.children.filter(c => !c.children).map(c => c.name));
    let dirNames = new Set(tree.children.filter(c => c.children).map(c => c.name + '.md'));

    let content = tree.children
      .filter(c => (!fileNames.has(c.name) || !dirNames.has(c.name)) && c.name != 'README.md')
      .map(c => this.renderToMd(c, dirNames.has(c.name + '.md') && fileNames.has(c.name + '.md')))
      .join('\n')
      .split('\n')
      .map(item => '  ' + item)
      .join('\n');
    let prefix = '';
    if (tree.name) {
      if (linkDir || fileNames.has('README.md')) {
        let linkPath = tree.link.replace(/ /g,'%20');
        if (fileNames.has('README.md')) linkPath += '/README.md';
        prefix = `- [${this.niceName(path.basename(tree.name, '.md'))}](${linkPath})\n`;
      }
      else prefix = `- ${this.niceName(tree.name)}\n`;
    }

    return prefix + content;
  }
}

index(path: string, leftPad: number) {
  let padding = new Array(leftPad + 1).join(' ')
  let root = this.buildTree(path);
  return this.renderToMd(root).split("\n").map(i => i.substr(2)).join('\n' + padding)
}

}


