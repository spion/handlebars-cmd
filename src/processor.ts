import { TOC as FileTree } from './toc';
import * as hbs from 'handlebars';
import * as path from 'path';
import * as fs from 'fs';
import * as glob from 'glob';
import * as gm from 'gray-matter';

let wd = process.cwd(),
  rootwd: string = null!;

hbs.registerHelper('include', function(this: any, file, opts: hbs.HelperOptions) {
  let ctx = null == opts.hash['context'] ? this : opts.hash['context'];
  let target = path.resolve(wd, file);
  let oldwd = wd;
  wd = path.dirname(target);
  let res = subHandle(target, ctx);
  wd = oldwd;
  return res;
});

hbs.registerHelper('filetree', function(dir: string, pad: number, base: string) {
  if (typeof pad !== 'number') pad = 0;
  let p = path.resolve(wd, dir);
  let relDir = path.relative(wd, p);
  if (typeof base !== 'string') {
    if (p !== dir && relDir.length > 0) base = '/' + relDir;
    else base = '';
  }
  let toc = new FileTree();
  return toc.index(p, pad, base);
});

hbs.registerHelper('glob', function(this: unknown, globString: string, opts: hbs.HelperOptions) {
  let localWD = opts.hash['wd'] ? path.resolve(wd, opts.hash['wd']) : wd;
  let items = glob.sync(globString, {
    cwd: localWD,
  });
  return items.map(item => opts.fn(this, { data: { file: item } })).join('');
});

hbs.registerHelper('cwd', function(this: unknown, dir: string, opts: hbs.HelperOptions) {
  let oldwd = wd;
  wd = path.resolve(wd, dir);
  let res = opts.fn(this);
  wd = oldwd;
  return res;
});

hbs.registerHelper('frontmatter', function(f: string, fieldName: string, opts: hbs.HelperOptions) {
  let filePath = path.resolve(wd, f);
  if (!fs.existsSync(filePath))
    throw new Error('Frontmatter: file ' + filePath + ' does not exist');
  let frontmatter = gm.read(filePath);
  return frontmatter.data[fieldName] || opts.hash['default'];
});

hbs.registerHelper('basename', function(filename, ext = '') {
  return path.basename(filename, ext);
});

hbs.registerHelper('file-exists', function(filename) {
  return fs.existsSync(path.resolve(wd, filename));
});

hbs.registerHelper('concat', function(...args) {
  return args.slice(0, -1).join('');
});

hbs.registerHelper('equals', function(arg1, arg2) {
  return arg1 == arg2;
});

hbs.registerHelper('left-pad', function(amount: number, content: string) {
  let spaces = new Array(amount + 1).join(' ');
  return content.split('\n').join('\n' + spaces);
});

hbs.registerHelper('cwd', function() {
  console.log(rootwd, wd);
  return path.relative(rootwd, wd);
});

function subHandle(fullPath: string, args: any) {
  wd = path.dirname(fullPath);
  let tmpl = fs.readFileSync(fullPath, 'utf8');

  var template = hbs.compile(tmpl.toString());
  var result = template(args);
  return result;
}

export function handle(fullPath: string, args: any) {
  rootwd = path.dirname(fullPath);
  return subHandle(fullPath, args);
}
