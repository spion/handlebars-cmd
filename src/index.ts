#!/usr/bin/env node

import * as fs from 'fs'
import * as yargs from 'yargs'
import * as hbs from 'handlebars'
import * as path from 'path'

import { handle } from './processor';

export let args = yargs
.usage('Usage: markdownbars -i input.hbs.md [-o output.md]')
.options({
  input: {
    alias: 'i',
    describe: 'Input file',
    type: 'string',
    demandOption: true
  },
  output: {
    alias: 'o',
    describe: 'Output file',
    type: 'string',
    demandOption: false
  },
  help: {
    alias: 'h',
    describe: 'Generate help',
    type: 'boolean',
    default: false
  },
  data: {
    alias: 'd',
    describe: 'Additional JSON data to pass into the template',
    type: 'string',
    default: '{}'
  }
}).argv;


let data = JSON.parse(args.data)
let filePath = path.resolve(process.cwd(), args.input);
let output = handle(filePath, data);
if (args.output) fs.writeFileSync(args.output, output)
else console.log(output);
