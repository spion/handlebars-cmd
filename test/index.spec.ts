import 'jest';
import { handle } from '../src/processor';

describe('markdownbars', () => {
  it('has working filetree', () => {
    let res = handle(__dirname + '/fixture/toc.tpl.md', {});
    expect(res).toMatchSnapshot();
  });
  it('has working glob', () => {
    let res = handle(__dirname + '/fixture/glob.tpl.md', {});
    expect(res).toMatchSnapshot();
  });

  it('supports complex example', () => {
    let res = handle(__dirname + '/complex-fixture/sidebar.tpl.md', {});
    expect(res).toMatchSnapshot();
  });
});
