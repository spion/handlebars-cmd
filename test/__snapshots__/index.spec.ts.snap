// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markdownbars has working filetree 1`] = `
"Things

- thingy
- - a
    - [x](/a/x.md)
    - [z](/a/z.md)
  - b
    - [y](/b/y.md)
  - [glob.tpl](/glob.tpl.md)
  - [toc.tpl](/toc.tpl.md)
- another
- - [x](/a/x.md)
  - [z](/a/z.md)
- - [x](/third-argument/x.md)
  - [z](/third-argument/z.md)
"
`;

exports[`markdownbars has working glob 1`] = `
"Sample glob:
  Path: a/x.md
  Path: a/z.md

  Path: a/x.md
  Path: a/z.md

  Path: This is X
  Path: z
"
`;

exports[`markdownbars supports complex example 1`] = `
"- f1
  - Guides
    - [First](packages/f1/test.md)
  - API
    
- [f2](packages/f2/README.md)
  - [Other thing](packages/f2/subdir/other.md)
  - [test](packages/f2/test.md)
"
`;
